### ISSUE TITLE
cuda set to 0 but still returns RuntimeError

### ISSUE THREAD

**Issue by enod** [allenai/allennlp#2225]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.5
 - AllenNLP version: 0.7.2
 - PyTorch version: 0.4.1

**Question**
I'm following the [tutorial](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/predicting_paper_venues/predicting_paper_venues_pt1.md) and customized a bit. Below is my configuration and runtime error.

```
{
  "dataset_reader": {
    "type": "yahoo_csv_reader",
    "lazy": true,
    "review": 0,
    "label": 1,
    "skip_header": true,
    "token_indexers": {
      "tokens": {
        "type": "single_id",
        "lowercase_tokens": true
    },
      "elmo": {
      "type": "elmo_characters"
    }}
  },
  "train_data_path": "tests/fixtures/kim_train.csv",
  "validation_data_path": "tests/fixtures/kim_valid.csv",
  "test_data_path": "tests/fixtures/kim_test.csv",
  "model": {
    "type": "review_classifier",
    "text_field_embedder": {
      "tokens": {
        "type": "embedding",
        "embedding_dim": 300,
        "pretrained_file": "https://s3-us-west-1.amazonaws.com/fasttext-vectors/word-vectors-v2/cc.ja.300.vec.gz",
        "trainable": true
      },
      "elmo": {
        "type": "elmo_token_embedder",
        "options_file": "elmo_trained/options.json",
        "weight_file": "elmo_trained/weights.hdf5",
        "do_layer_norm": false,
        "dropout": 0.5
      }
    },
    "review_encoder": {
      "type": "boe",
      "embedding_dim": 1324,
      "averaged": true
    },
    "classifier_feedforward": {
      "input_dim": 1324,
      "num_layers": 2,
      "hidden_dims": [200, 3],
      "activations": ["relu", "linear"],
      "dropout": [0.2, 0.0]
    }
  },
  "iterator": {
    "type": "bucket",
    "sorting_keys": [["review", "num_tokens"]],
    "padding_noise": 0.0,
    "batch_size": 32
  },
  "trainer": {
    "num_epochs": 10,
    "patience": 5,
    "cuda_device": 0,
    "grad_clipping": 5.0,
    "validation_metric": "+accuracy",
    "optimizer": {
      "type": "adagrad"
    }
  }
}
```

As you can see I set cuda_device to 0 in order to utilize GPU. When I run it:

```
Traceback (most recent call last):
  File "/mnt/data/elmo/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/mnt/data/elmo/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/run.py", line 18, in <module>
    main(prog="allennlp")
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/commands/__init__.py", line 72, in main
    args.func(args)
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/commands/train.py", line 111, in train_model_from_args
    args.force)
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/commands/train.py", line 142, in train_model_from_file
    return train_model(params, serialization_dir, file_friendly_logging, recover, force)
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/commands/train.py", line 351, in train_model
    metrics = trainer.train()
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/training/trainer.py", line 755, in train
    train_metrics = self._train_epoch(epoch)
  File "/mnt/data/elmo/lib/python3.6/site-packages/allennlp/training/trainer.py", line 525, in _train_epoch
    self.optimizer.step()
  File "/mnt/data/elmo/lib/python3.6/site-packages/torch/optim/adagrad.py", line 92, in step
    state['sum'].addcmul_(1, grad, grad)
RuntimeError: Expected object of type torch.FloatTensor but found type torch.cuda.FloatTensor for argument #4 'tensor1'
[INFO/MainProcess] process shutting down
```
It it expecting CPU tensor not the cuda tensor. But doesn't setting gpu_device=0 handles it?

**Comment by joelgrus** [comment]
this is possibly this bug:

https://github.com/allenai/allennlp/issues/2189

it should be fixed in the latest release 0.8 (and also should not happen if you use a different optimizer)

**Comment by enod** [comment]
changed to adam and it's working. Thanks!
